src/main/java/com/epam/aidial/core/storage/BlobWriteStream.java

```diff
@@ -0,0 +1,187 @@
+package com.epam.aidial.core.storage;
+
+import com.epam.aidial.core.data.FileMetadata;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.streams.WriteStream;
+import lombok.extern.slf4j.Slf4j;
+import org.jclouds.blobstore.domain.MultipartPart;
+import org.jclouds.blobstore.domain.MultipartUpload;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Implementation of vertx {@link io.vertx.core.streams.WriteStream} that handles data chunks (from {@link io.vertx.core.streams.ReadStream}) and writes them to the blob storage.
+ * If file content is bigger than 5MB - multipart upload will be used.
+ * Chunk size can be configured via {@link #setWriteQueueMaxSize(int)} method, but should be no less than 5 MB according to the s3 specification.
+ * If any exception is caught in between - multipart upload will be aborted.
+ */
+@Slf4j
+public class BlobWriteStream implements WriteStream<Buffer> {
```

Comment:
We should try to write Unit test simulating slow upload to blob storage. The blob storage could be mocked.

Reply:
let's add in a separate PR